authorizer-app-backend/src/main/java/org/radarbase/authorizer/service/OAuth1RestSourceAuthorizationService.kt

```diff
@@ -79,111 +94,112 @@ abstract class OAuth1RestSourceAuthorizationService(
+                tokens.tokenSecret?.let { parameters.append(OAUTH_ACCESS_TOKEN_SECRET, it) }
+            }
+            parameters.append(
+                OAUTH_CALLBACK,
+                URLBuilder(config.service.callbackUrl).run {
+                    parameters.append("state", state)
+                    buildString()
+                },
+            )
+            buildString()
+        }
     }
 
-    private fun requestToken(
+    private suspend fun requestToken(
         tokenEndpoint: String?,
         tokens: RestOauth1AccessToken,
         sourceType: String,
-    ): RestOauth1AccessToken? {
-        val req = createRequest(
-            "POST",
+    ): RestOauth1AccessToken? = withContext(Dispatchers.IO) {
+        val response = request(
+            HttpMethod.Post,
             tokenEndpoint.orEmpty(),
             tokens,
             sourceType,
         )
 
-        return httpClient.newCall(req)
-            .execute()
-            .use { response ->
-                when (response.code) {
-                    200 -> response.body?.string()
-                        ?.let { tokenReader.readValue<RestOauth1AccessToken>(parseParams(it)) }
-                        ?: throw HttpBadGatewayException("Service did not provide a result")
-                    400, 401, 403 -> null
-                    else -> throw HttpBadGatewayException("Cannot connect to $tokenEndpoint: HTTP status ${response.code}")
-                }
+        when (response.status) {
+            HttpStatusCode.OK -> try {
+                Json.decodeFromString(
+                    RestOauth1AccessToken.serializer(),
+                    response.bodyAsText().toJsonString(),
+                )
+            } catch (ex: IllegalArgumentException) {
+                throw HttpBadGatewayException("Service did not provide a result: $ex")
             }
+            HttpStatusCode.BadRequest, HttpStatusCode.Unauthorized, HttpStatusCode.Forbidden -> null
+            else -> throw HttpBadGatewayException("Cannot connect to ${response.request.url}: HTTP status ${response.status}")
+        }
     }
 
-    fun createRequest(method: String, url: String, tokens: RestOauth1AccessToken, sourceType: String): Request {
+    suspend fun request(method: HttpMethod, url: String, tokens: RestOauth1AccessToken, sourceType: String): HttpResponse {
         val authConfig = clientService.forSourceType(sourceType)
         val params = this.getAuthParams(authConfig, tokens.token, tokens.tokenVerifier)
         params[OAUTH_SIGNATURE] =
             OauthSignature(url, params, method, authConfig.clientSecret, tokens.tokenSecret).getEncodedSignature()
-        val headers = params.toFormattedHeader()
 
-        return Request.Builder()
-            .url(url)
-            .header("Authorization", "OAuth $headers")
-            .method(method, if (method == "POST") "".toRequestBody(null) else null)
-            .build()
+        return httpClient.request(url = Url(url)) {
```

Comment:
If we run `withContext` here, and remove it from the methods that call `request()`, will those still run in the context? Or would we need to still run `withContext` in the outer methods?

Reply:
yes, you're right, I've added it to the lowermost function now. Indeed `withContext` calls stack without problem. If it's already in the right context it won't do any switches.